### ISSUE TITLE
Can't build on Windows (Visual Studio), what am I doing wrong?

### ISSUE THREAD

**Issue by tomq42** [grpc/grpc#6335]
I'm struggling to build on Windows.
I've done a git clone (e107f8e6884b2f8edec17869bbe0d42f35e88445 if it makes a difference), and then done a git submodule update. I've also done a nuget restore.

I get a number of compile errors in boringssl, missing header files. For example:
grpc\third_party\boringssl\ssl\t1_lib.c(117): fatal error C1083: Cannot open include file: 'openssl/bytestring.h': No such file or directory

This file, openssl/bytestring.h does indeed not exist in the nuget downloaded package in grpc\vsprojects\packages\grpc.dependencies.openssl.1.0.204.1\build\native\include, however it DOES exist in the git submodule in third_party\boringssl\include\openssl.

So the issue appears to either be that I've ended up with the wrong nuget package, or that the nuget package is itself wrong?

Thanks.

**Comment by tomq42** [comment]
In case if wasn't clear the above was refering to me attempting to build using Visual Studio projects.

However I've tried using msys and that doesn't work either.

**Comment by nicolasnoble** [comment]
You're not saying what you're trying to build exactly. BUT from your error
message, I'm guessing you went for "Build the Solution", which isn't the
proper way. If you want to build the grpc library using Visual Studio, open
grpc.sl, select the grpc project, then build it. If you want to look at the
tests and samples, open buildtests_c.sln, and build individual tests. Don't
try to build the whole solution.

For msys, without any further error detail, it'd be hard to help.

**Comment by tomq42** [comment]
I followed the instructions in the readme which say:
"After that, you can build the solution using one of these options: 1. open grpc.sln with Visual Studio and hit "Build"" . 2. build from commandline using msbuild grpc.sln /p:Configuration=Debug"

If I build each project individually though, then all build except for boringssl (with the file not found errors above) and z (with some errors
"error C4996: 'write': The POSIX name for this item is deprecated"
)

Which makes me wonder why they are in the solution. My suspicion is I'd need them when I actually tried to write anything that used either security or compression, in which case I'd be back to the same position.

Either way he experience is poor compared to other open source projects, which is a surprise.

**Comment by tomq42** [comment]
After a long and complicated sequence of steps I *think* I might have a built copy of GRPC. 
Still far from clear what the equivalent of "make install" is, that is, how I gather together the things I need to actually "install", what that list of files is. I'm assuming I need some headers and libs, and then some unknown pieces of protobuf, protoc compiler and protoc compiler plugin.

Given that it looks like there is a Windows build done on jenkins, and given the apparent complexity of building this correctly, perhaps consider building a zip bundle that can be pushed somewhere?

**Comment by MrMoose** [comment]
I fixed this by manually changing the msvc project files so they include the bundled boringssl solution. It is built but the rest of the solution never includes it. It's a bug in the project files. My fix is kinda shotgun though so I dare not commit it.

HTH,
Stephan

**Comment by Kurtpenter** [comment]
@tomq42 can you please add here the steps you did to make it build? As @MrMoose suggested I added the include:

> $(SolutionDir)\..\third_party\boringssl\include;

this solves the openssl missing includes but does not seem to be enough.

**Comment by tomq42** [comment]
So I've just taken a fresh clone, and synced the submodules.

If I just do a "build solution" at this point, I get a bunch of errors about openssl, and the "boringssl" and "z" projects fail to build.

So I remove those two projects, on the basis that they are provided through nuget. Solution now builds.
What I don't know, can't remember at the moment, is what else I did. I recall manually having to put protobuf into the thirdparty directory as per the windows readme. I also remember the step required to get to point I could run the protoc compiler with the grpc plugins being a bit complicated, but I can't remember how now.

So I *think* that to get the Visual Studio solution to build, you just remove "boringssl" and "z". However, just building the Visual Studio solution isn't enough.

Basically there needs to be a one shot script somewhere, equivalent to the unix "make install", as currently it's just too error prone.

**Comment by mosra** [comment]
As noted in #4778 and #5340, the solution is to invoke `msbuild` as following to use boringssl and zlib provided by nuget instead of building them from scratch:

```bat
msbuild grpc.sln /t:grpc
```

I spent an awful lot of time finding this workaround. Would it be possible for gRPC developers to add even a tiny note about this to the readme at https://github.com/grpc/grpc/blob/master/vsprojects/README.md so others don't have to dig it in comments of old closed issues?

Thanks.

cc: @lumpidu

**Comment by MrMoose** [comment]
Quite honestly, this wouldn't help much. Maintaining and checking in platform and tool specific solution files and build system feels last century to me and should not be done in this day and age anymore. Especially not in new projects. Get rid of all this. CMake exists. Make use of it.

Just my 2 cents...
Moose

**Comment by mosra** [comment]
Oh, yes, please. I'm maintaining my own CMake build scripts for gRPC just to be able to build the C++ (not ObjC) version for iOS. After some more angry fighting with sln files I ended up reusing CMake also for Windows.

**Comment by MrMoose** [comment]
Good call. Are they good enough for a commit? Maybe you can convince the maintainers to accept them... I suppose they should see in what state this build system is as well and would be glad to get some help with that. Even if not, you can still publish it in a fork.

**Comment by MrMoose** [comment]
Oops. I just saw there is already an issue to that: #1096

**Comment by mehrdada** [comment]
We are cleaning up the gRPC issue backlog and it sounds like the issue is resolved already. If you are still experiencing this issue, please re-open the issue with information on how to reproduce.
